api/v1alpha1/gateway_types.go

```diff
@@ -146,6 +150,40 @@ type ListenerTLS struct {
 	Options map[string]string `json:"options"`
 }
 
+// RouteSelector provides a mechanism to associate routes with this Gateway.
+type RouteSelector struct {
+	// Routes specifies what routes to associate with this Gateway. If defined,
+	// NamespaceSelector and LabelSelector are ignored and only the specified
+	// routes are associated with this Gateway.
+	//
+	// If undefined, all routes or routes matching NamespaceSelector and/or
+	// LabelSelector are associated with this Gateway.
+	//
+	// +optional
+	Routes []core.TypedLocalObjectReference `json:"routes"`
+	// NamespaceSelector specifies a set of namespace labels used for selecting
+	// routes to associate with this Gateway.
+	//
+	// If undefined, routes from all namespaces will be associated to this Gateway.
+	// If undefined and LabelSelector is defined, routes containing LabelSelector
+	// labels across all namespaces are associated to this Gateway. If NamespaceSelector
+	// and LabelSelector are defined, only routes matching the LabelSelector labels
+	// from NamespaceSelector namespaces are associated to this Gateway.
+	//
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
```

Comment:
This seems like an important difference, as you appear to be proposing the ability to admit routes to the gateway outside the gateway's namespace. My understanding of the current API is that only routes within the gateway's namespace can be admitted into that gateway. Is my understanding of the current API correct?

Reply:
Looking back over the the API sketch, Routes can be admitted to multiple gateways through route delegation and exporting, but it doesn't seem like namespaced references can be achieved directly with the Gateway->Route association. If true, I'm not sure if it's intentional.

The thing that caught my eye is the API this change is replacing seems to be scoped exclusively to local references, and the change would thus expand the scope to include namespaced references.

Reply:
@ironcladlou my understanding is that a gateway supports route binding from namespaces that differ from the gateway's namespace. See the gateway config [sample](https://github.com/kubernetes-sigs/service-apis/blob/master/config/samples/networking_v1alpha1_gateway.yaml#L30-L36) that binds routes from 2 different namespaces to gateway "gateway-sample". The namespaces of both routes differ from the gateway namespace.

Reply:
Yes -- at least in the sketch (and we should make it explict), the idea is that gateway <=> route /can/ cross namespaces, to meet the gateway admin/user app separation.

I will add this to the use cases doc

Reply:
@bowei 

Thanks for the clarification of intent. Looks like there's a possible mismatch here:

https://github.com/kubernetes-sigs/service-apis/blob/master/api/v1alpha1/gateway_types.go#L61

https://github.com/kubernetes-sigs/service-apis/blob/master/config/samples/networking_v1alpha1_gateway.yaml#L32

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#typedlocalobjectreference-v1-core

Looks like maybe `ObejctReference` would be appropriate instead of `TypedLocalObjectReference`?

Reply:
Commit `e0637e3` updated the `Route` type to `ObejctReference` and the associated crd.